neurom/core/morphology.py

```diff
@@ -295,21 +293,24 @@ def iter_neurites(
         last_position = max(NRN_ORDER.values()) + 1
         neurites = sorted(neurites, key=lambda neurite: NRN_ORDER.get(neurite.type, last_position))
 
-    if use_subtrees:
-        neurites = flatten(
-            _homogeneous_subtrees(neurite) if neurite.is_heterogeneous() else [neurite]
-            for neurite in neurites
-        )
-
     neurite_iter = iter(neurites) if filt is None else filter(filt, neurites)
 
     if mapfun is None:
         return neurite_iter
 
-    if use_subtrees:
-        return (mapfun(neurite, section_type=neurite.type) for neurite in neurite_iter)
-
-    return map(mapfun, neurite_iter)
+    return (
+        (
+            mapfun(
+                neurite,
+                section_type=filt.type
+                if (filt is not None or filt.type is NeuriteType.all)
+                else None,
+            )
+            if neurite.process_subtrees
+            else mapfun(neurite, section_type=NeuriteType.all)
```

Comment:
Does passing the `NeuriteType.all` forces the filter to go through all the sections?

Reply:
It should not change anything, it's just to have a consistent signature. If we don't do this we can't use lambdas as functions. But we could use inspection instead and pass a section type only when it is needed.